<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Dependencies

In [1]:
# !pip install -r requirements.txt
!pip freeze

anyio==4.3.0
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
arrow==1.3.0
asttokens==2.4.1
async-lru==2.0.4
attrs==23.2.0
Babel==2.14.0
beautifulsoup4==4.12.3
bleach==6.1.0
certifi==2024.2.2
cffi==1.16.0
charset-normalizer==3.3.2
colorama==0.4.6
comm==0.2.1
debugpy==1.8.1
decorator==5.1.1
defusedxml==0.7.1
executing==2.0.1
fastjsonschema==2.19.1
fqdn==1.5.1
greenlet==3.0.3
h11==0.14.0
httpcore==1.0.4
httpx==0.27.0
idna==3.6
ipykernel==6.29.3
ipython==8.22.2
ipython-genutils==0.2.0
ipython-sql==0.5.0
isoduration==20.11.0
jedi==0.19.1
Jinja2==3.1.3
json5==0.9.22
jsonpointer==2.4
jsonschema==4.21.1
jsonschema-specifications==2023.12.1
jupyter-events==0.9.0
jupyter-lsp==2.2.4
jupyter_client==8.6.1
jupyter_core==5.7.2
jupyter_server==2.13.0
jupyter_server_terminals==0.5.2
jupyterlab==4.1.4
jupyterlab-vim==4.1.3
jupyterlab_pygments==0.3.0
jupyterlab_server==2.25.4
MarkupSafe==2.1.5
matplotlib-inline==0.1.6
mistune==3.0.2
nbclient==0.9.0
nbconvert==7.16.2
nbformat==5.9.2
nest-asyncio==1.6.0


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [2]:
import pandas as pd
import sqlite3

In [3]:
conn = sqlite3.connect('FinalDB.db')

Load the SQL magic module


In [4]:
%load_ext sql

Establish a connection between SQL magic module and the database `FinalDB.db`


In [5]:
%sql sqlite:///FinalDB.db

In [6]:
%%sql

drop table if exists CENSUS;
drop table if exists PUBLIC_SCHOOL;
drop table if exists CRIME;

 * sqlite:///FinalDB.db
Done.
Done.
Done.


[]

Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [7]:
census_data = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01')
public_schools_data = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01')
crime_data = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01')

census_data.to_sql('CENSUS', conn)
public_schools_data.to_sql('PUBLIC_SCHOOL', conn)
crime_data.to_sql('CRIME', conn)

533

You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


In [8]:
%%sql

select count(*)
from CRIME;

 * sqlite:///FinalDB.db
Done.


count(*)
533


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [9]:
%%sql
SELECT * FROM sqlite_master WHERE type='table';

 * sqlite:///FinalDB.db
Done.


type,name,tbl_name,rootpage,sql
table,CENSUS,CENSUS,2,"CREATE TABLE ""CENSUS"" (""index"" INTEGER, ""COMMUNITY_AREA_NUMBER"" REAL, ""COMMUNITY_AREA_NAME"" TEXT, ""PERCENT_OF_HOUSING_CROWDED"" REAL, ""PERCENT_HOUSEHOLDS_BELOW_POVERTY"" REAL, ""PERCENT_AGED_16__UNEMPLOYED"" REAL, ""PERCENT_AGED_25__WITHOUT_HIGH_SCHOOL_DIPLOMA"" REAL, ""PERCENT_AGED_UNDER_18_OR_OVER_64"" REAL, ""PER_CAPITA_INCOME"" INTEGER, ""HARDSHIP_INDEX"" REAL)"
table,PUBLIC_SCHOOL,PUBLIC_SCHOOL,6,"CREATE TABLE ""PUBLIC_SCHOOL"" (""index"" INTEGER, ""School_ID"" INTEGER, ""NAME_OF_SCHOOL"" TEXT, ""Elementary, Middle, or High School"" TEXT, ""Street_Address"" TEXT, ""City"" TEXT, ""State"" TEXT, ""ZIP_Code"" INTEGER, ""Phone_Number"" TEXT, ""Link"" TEXT, ""Network_Manager"" TEXT, ""Collaborative_Name"" TEXT, ""Adequate_Yearly_Progress_Made_"" TEXT, ""Track_Schedule"" TEXT, ""CPS_Performance_Policy_Status"" TEXT, ""CPS_Performance_Policy_Level"" TEXT, ""HEALTHY_SCHOOL_CERTIFIED"" TEXT, ""Safety_Icon"" TEXT, ""SAFETY_SCORE"" REAL, ""Family_Involvement_Icon"" TEXT, ""Family_Involvement_Score"" TEXT, ""Environment_Icon"" TEXT, ""Environment_Score"" REAL, ""Instruction_Icon"" TEXT, ""Instruction_Score"" REAL, ""Leaders_Icon"" TEXT, ""Leaders_Score"" TEXT, ""Teachers_Icon"" TEXT, ""Teachers_Score"" TEXT, ""Parent_Engagement_Icon"" TEXT, ""Parent_Engagement_Score"" TEXT, ""Parent_Environment_Icon"" TEXT, ""Parent_Environment_Score"" TEXT, ""AVERAGE_STUDENT_ATTENDANCE"" TEXT, ""Rate_of_Misconducts__per_100_students_"" REAL, ""Average_Teacher_Attendance"" TEXT, ""Individualized_Education_Program_Compliance_Rate"" TEXT, ""Pk_2_Literacy__"" TEXT, ""Pk_2_Math__"" TEXT, ""Gr3_5_Grade_Level_Math__"" TEXT, ""Gr3_5_Grade_Level_Read__"" TEXT, ""Gr3_5_Keep_Pace_Read__"" TEXT, ""Gr3_5_Keep_Pace_Math__"" TEXT, ""Gr6_8_Grade_Level_Math__"" TEXT, ""Gr6_8_Grade_Level_Read__"" TEXT, ""Gr6_8_Keep_Pace_Math_"" TEXT, ""Gr6_8_Keep_Pace_Read__"" TEXT, ""Gr_8_Explore_Math__"" TEXT, ""Gr_8_Explore_Read__"" TEXT, ""ISAT_Exceeding_Math__"" REAL, ""ISAT_Exceeding_Reading__"" REAL, ""ISAT_Value_Add_Math"" REAL, ""ISAT_Value_Add_Read"" REAL, ""ISAT_Value_Add_Color_Math"" TEXT, ""ISAT_Value_Add_Color_Read"" TEXT, ""Students_Taking__Algebra__"" TEXT, ""Students_Passing__Algebra__"" TEXT, ""9th Grade EXPLORE (2009)"" TEXT, ""9th Grade EXPLORE (2010)"" TEXT, ""10th Grade PLAN (2009)"" TEXT, ""10th Grade PLAN (2010)"" TEXT, ""Net_Change_EXPLORE_and_PLAN"" TEXT, ""11th Grade Average ACT (2011)"" TEXT, ""Net_Change_PLAN_and_ACT"" TEXT, ""College_Eligibility__"" TEXT, ""Graduation_Rate__"" TEXT, ""College_Enrollment_Rate__"" TEXT, ""COLLEGE_ENROLLMENT"" INTEGER, ""General_Services_Route"" INTEGER, ""Freshman_on_Track_Rate__"" TEXT, ""X_COORDINATE"" REAL, ""Y_COORDINATE"" REAL, ""Latitude"" REAL, ""Longitude"" REAL, ""COMMUNITY_AREA_NUMBER"" INTEGER, ""COMMUNITY_AREA_NAME"" TEXT, ""Ward"" INTEGER, ""Police_District"" INTEGER, ""Location"" TEXT)"
table,CRIME,CRIME,11,"CREATE TABLE ""CRIME"" (""index"" INTEGER, ""ID"" INTEGER, ""CASE_NUMBER"" TEXT, ""DATE"" TEXT, ""BLOCK"" TEXT, ""IUCR"" TEXT, ""PRIMARY_TYPE"" TEXT, ""DESCRIPTION"" TEXT, ""LOCATION_DESCRIPTION"" TEXT, ""ARREST"" INTEGER, ""DOMESTIC"" INTEGER, ""BEAT"" INTEGER, ""DISTRICT"" INTEGER, ""WARD"" REAL, ""COMMUNITY_AREA_NUMBER"" REAL, ""FBICODE"" TEXT, ""X_COORDINATE"" REAL, ""Y_COORDINATE"" REAL, ""YEAR"" INTEGER, ""LATITUDE"" REAL, ""LONGITUDE"" REAL, ""LOCATION"" TEXT)"


### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [10]:
%%sql

select COMMUNITY_AREA_NUMBER, COMMUNITY_AREA_NAME
from CENSUS
where PER_CAPITA_INCOME < 11000;

 * sqlite:///FinalDB.db
Done.


COMMUNITY_AREA_NUMBER,COMMUNITY_AREA_NAME
26.0,West Garfield Park
30.0,South Lawndale
37.0,Fuller Park
54.0,Riverdale


### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [11]:
%%sql

select CASE_NUMBER
from CRIME
where DESCRIPTION LIKE '%MINOR%';

 * sqlite:///FinalDB.db
Done.


CASE_NUMBER
HL266884
HK238408


### Problem 4

##### List all kidnapping crimes involving a child?


In [12]:
%%sql

select *
from CRIME
where PRIMARY_TYPE like '%kidnap%' and DESCRIPTION like '%child%';

 * sqlite:///FinalDB.db
Done.


index,ID,CASE_NUMBER,DATE,BLOCK,IUCR,PRIMARY_TYPE,DESCRIPTION,LOCATION_DESCRIPTION,ARREST,DOMESTIC,BEAT,DISTRICT,WARD,COMMUNITY_AREA_NUMBER,FBICODE,X_COORDINATE,Y_COORDINATE,YEAR,LATITUDE,LONGITUDE,LOCATION
520,5276766,HN144152,2007-01-26,050XX W VAN BUREN ST,1792,KIDNAPPING,CHILD ABDUCTION/STRANGER,STREET,0,0,1533,15,29.0,25.0,20,1143050.0,1897546.0,2007,41.87490841,-87.75024931,"(41.874908413, -87.750249307)"


### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


In [13]:
%%sql

select distinct(PRIMARY_TYPE)
from CRIME
where LOCATION_DESCRIPTION like '%school%'
limit 20;

 * sqlite:///FinalDB.db
Done.


PRIMARY_TYPE
BATTERY
CRIMINAL DAMAGE
NARCOTICS
ASSAULT
CRIMINAL TRESPASS
PUBLIC PEACE VIOLATION


### Problem 6

##### List the type of schools along with the average safety score for each type.


In [14]:
%%sql

select `Elementary, Middle, or High School` as S, avg(SAFETY_SCORE)
from PUBLIC_SCHOOL
group by S
limit 20;

 * sqlite:///FinalDB.db
Done.


S,avg(SAFETY_SCORE)
ES,49.52038369304557
HS,49.62352941176471
MS,48.0


### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [15]:
%%sql

select COMMUNITY_AREA_NAME, PERCENT_HOUSEHOLDS_BELOW_POVERTY
from CENSUS
order by PERCENT_HOUSEHOLDS_BELOW_POVERTY desc
limit 5;

 * sqlite:///FinalDB.db
Done.


COMMUNITY_AREA_NAME,PERCENT_HOUSEHOLDS_BELOW_POVERTY
Riverdale,56.5
Fuller Park,51.2
Englewood,46.6
North Lawndale,43.1
East Garfield Park,42.4


### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [16]:
%%sql

select COMMUNITY_AREA_NUMBER
from (
    select COMMUNITY_AREA_NUMBER, count(COMMUNITY_AREA_NUMBER) as Frequency
    from CRIME
    group by COMMUNITY_AREA_NUMBER
    order by Frequency desc
    limit 1
);

 * sqlite:///FinalDB.db
Done.


COMMUNITY_AREA_NUMBER
25.0



Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [17]:
%%sql

select COMMUNITY_AREA_NAME
from (select * from CENSUS order by HARDSHIP_INDEX desc limit 1);

 * sqlite:///FinalDB.db
Done.


COMMUNITY_AREA_NAME
Riverdale


### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [18]:
%%sql

select COMMUNITY_AREA_NAME
from (select COMMUNITY_AREA_NUMBER, count(COMMUNITY_AREA_NUMBER) as Crimes from CRIME group by COMMUNITY_AREA_NUMBER) as A
join CENSUS as B on A.COMMUNITY_AREA_NUMBER=B.COMMUNITY_AREA_NUMBER
order by A.Crimes desc
limit 1;

 * sqlite:///FinalDB.db
Done.


COMMUNITY_AREA_NAME
Austin


## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>

## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |

## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
